### PR TITLE
Feature/SBPS value-selection heuristic

### PR DESCRIPTION
**Pull request by raphaelboudreault**
This PR implements a new search option in Chuffed for optimization problems:

`--sbps on` tells the solver to use _Solution-based phase saving_ (SBPS) value selection heuristic for all variables (eager integers, lazy integers, SAT/boolean).

SBPS is based on the following paper:

> E. Demirović, G. Chu, and P. J. Stuckey. Solution-Based Phase Saving for CP: A Value-Selection Heuristic to Simulate Local Search Behavior in Complete Solvers. In *Principles and Practice of Constraint Programming*, Cham, 2018, p. 99‑108.

SBPS, used with restarts and an activity-based variable selection (VSIDS), simulates an automatic *Large Neighborhood Search* in the solver.  When branching on a variable, it branches if possible on the value this variable had in the best solution so far. If it is not possible, the value selection heuristic used is the user-defined one. In the context of a RCPSP I've worked on with an industrial partner, under the supervision of Claude-Guy Quimper, usage of this heuristic led to significant improvements, finding better solutions faster. This implementation is based on the original code of the related paper (we greatly thank Emir Demirović for that!), which have been updated for compatibility reasons with lazy integer variables.

### CODE REVIEW

File: chuffed/vars/int-var.cpp [hunk 1]
```diff
@@ -30,6 +30,8 @@ IntVar::IntVar(int _min, int _max) :
   , preferred_val(PV_MIN)
   , activity(0)
   , in_queue(false)
+    , SBPSValueSelection(false)
```
Comment (schutta) [inline chuffed/vars/int-var.cpp]: Please rename it and all its occurrences to "sbps_value_selection" to be consistent with existing naming convention.

File: chuffed/vars/int-var.h [hunk 2]
```diff
@@ -120,6 +120,14 @@ class IntVar : public Var {
 	void setPreferredVal(PreferredVal p) { preferred_val = p; }
 	DecInfo* branch();
 
+    // Solution-based phase saving
+    bool SBPSValueSelection;
```
Comment (schutta) [inline chuffed/vars/int-var.h]: Please rename it and all its occurrences to "sbps_value_selection" to be consistent with existing naming convention.

File: chuffed/vars/int-var.h [hunk 3]
```diff
@@ -120,6 +120,14 @@ class IntVar : public Var {
...
+    int lastSolutionValue;
```
Comment (schutta) [inline chuffed/vars/int-var.h]: Please rename it and all its occurrences to "last_solution_value" to be consistent with existing naming convention.

File: chuffed/vars/int-var.cpp [hunk 4]
```diff
@@ -30,6 +30,8 @@ IntVar::IntVar(int _min, int _max) :
...
+    , lastSolutionValue(-1)
```
Comment (schutta) [inline chuffed/vars/int-var.cpp]: Please rename it and all its occurrences to "last_solution_value" to be consistent with existing naming convention.

**Comment by schutta** [comment]
Please perform the adjustments mentioned above, so that we can finalise the merge request.

**Comment by raphaelboudreault** [comment]
The adjustments have been made. Thank you.

**Review by schutta — APPROVED** [review]

**Pull request merged**